angelsrefining/prototypes/angels-functions.lua

```diff
@@ -1375,17 +1375,54 @@ function angelsmods.functions.remove_crafting_category(crafting_machine_type, cr
   end
 end
 
+-------------------------------------------------------------------------------
+-- MODIFY FAST_REPLACE_CATEGORY -----------------------------------------------
+-------------------------------------------------------------------------------
+function angelsmods.functions.set_fast_replace_category(crafting_machine_type, crafting_machine_name, next_upgrade)
+  --search for new category (if needed), skip if identical
+  if not data.raw[crafting_machine_type] then return end
+  local crafting_machine1 = data.raw[crafting_machine_type][crafting_machine_name]
+  local crafting_machine2 = data.raw[crafting_machine_type][next_upgrade]
+  if not crafting_machine1 then return end
+  if not crafting_machine2 then return end
+  --get current FRC
+  local FRC1 = crafting_machine1.fast_replaceable_group
+  local FRC2 = crafting_machine2.fast_replaceable_group
+
+  if FRC1 == FRC2 then
+    --no need to change
+  elseif FRC2 ~= nil then --change it
+    crafting_machine1.fast_replaceable_group = FRC2
+    if crafting_machine1.collision_box[1][1] ~= crafting_machine2.collision_box[1][1] then
+      if crafting_machine1.collision_box[2][2] ~= crafting_machine2.collision_box[2][2] then
+        --boxes don't match... nil out the properties
+        crafting_machine1.fast_replaceable_group = nil
+        --clobber next_upgrade too
+        crafting_machine1.next_upgrade = nil
+      else --set fast_replace group
+        if FRC ~= nil then
+          crafting_machine1.fast_replaceable_group = FRC2
+        end
+      end
+    end

```

Comment:
I'm not following this a 100%, shouldn't you check all 4 values, so `left_top.x`, `left_top.y`, `right_bottom.x` and `right_bottom.y`? So if they all match, you can check the FRC, otherwise you set it to nil

```lua
local function box_equal(b1, b2)
  if not (b1 and b2) then return false end

  local function pos_equal(p1, p2)
    if not (p1 and p2) then return false end

    local p1x = p1.x or p1[1] or nil
    local p2x = p2.x or p2[1] or nil
    if not (p1x and p1x == p2x) then return false end

    local p1y = p1.y or p1[2] or nil
    local p2y = p2.y or p2[2] or nil

    return (p1y and p1y == p2y)
  end

  if not pos_equal(b1.left_top or b1[1], b2.left_top or b2[1]) then return false end
  return pos_equal(b1.right_bottom or b1[2], b2.right_bottom or b2[2])
end
```

such that
```lua
if box_equal(crafting_machine1.collision_box, crafting_machine2.collision_box) then
  -- equal logic (see next comment)
else
  --boxes don't match... nil out the properties
  crafting_machine1.fast_replaceable_group = nil
  --clobber next_upgrade too
  crafting_machine1.next_upgrade = nil
end

Reply:
unless something maddening is happening, the items are all rectangular, so top left and bottom right should be fine?

Reply:
But why are you only checking [1][1] aka left_top.x and [2][2] aka right_bottom.y? what about left_top.y and right_bottom.x?

Reply:
oh, good point, derp